File: promo_codes.py
#!/usr/bin/env python3
"""Hotel Tower promo code generator / validator.

Mirrors the algorithm in HotelTower/HotelTower.ino: codes are hu-{hash},
8 chars from a legible alphabet (no 0/1/i/l/o), derived from the floor
count with splitmix64 and a fixed salt. One code per floor >= 50.

Usage:
  python3 promo_codes.py            # print codes for floors 50..100
  python3 promo_codes.py 50 200     # print codes for floors 50..200
"""
import sys

MASK = (1 << 64) - 1
ALPHA = "23456789abcdefghjkmnpqrstuvwxyz"  # 31 legible chars


def promo_code(floors: int) -> str:
    x = ((floors * 0x9E3779B97F4A7C15) ^ 0x48554E4956455253) & MASK
    x ^= x >> 30
    x = (x * 0xBF58476D1CE4E5B9) & MASK
    x ^= x >> 27
    x = (x * 0x94D049BB133111EB) & MASK
    x ^= x >> 31
    h = ""
    for _ in range(8):
        h += ALPHA[x % 31]
        x //= 31
    return f"hu-{h}"


if __name__ == "__main__":
    lo = int(sys.argv[1]) if len(sys.argv) > 1 else 50
    hi = int(sys.argv[2]) if len(sys.argv) > 2 else 100
    for f in range(lo, hi + 1):
        print(f"{f}\t{promo_code(f)}")

File: promo_codes.js
// Hotel Tower promo code validation / parsing.
// Mirrors HotelTower/HotelTower.ino (and promo_codes.py): codes are
// "hu-" + 8 chars from a legible alphabet (no 0/1/i/l/o), derived from the
// floor count via splitmix64 with a fixed salt. One unique code per floor.
//
// Usage (ESM):
//   import { parsePromoCode } from "./promo_codes.js";
//   parsePromoCode("hu-nbem7c4y");  // -> { valid: true, floors: 50 }
//   parsePromoCode("hu-nope1234"); // -> { valid: false, floors: null }

const MASK = (1n << 64n) - 1n;
const ALPHA = "23456789abcdefghjkmnpqrstuvwxyz"; // 31 legible chars

const MIN_FLOORS = 50;  // first code is earned at floor 50
const MAX_FLOORS = 400; // the game's hard floor cap (kMaxFloors)

/** Compute the code for a floor count, e.g. 50 -> "hu-nbem7c4y". */
export function promoCode(floors) {
  let x =
    ((BigInt(floors) * 0x9e3779b97f4a7c15n) ^ 0x48554e4956455253n) & MASK;
  x ^= x >> 30n;
  x = (x * 0xbf58476d1ce4e5b9n) & MASK;
  x ^= x >> 27n;
  x = (x * 0x94d049bb133111ebn) & MASK;
  x ^= x >> 31n;
  let h = "";
  for (let i = 0; i < 8; i++) {
    h += ALPHA[Number(x % 31n)];
    x /= 31n;
  }
  return "hu-" + h;
}

// code -> floors, built once
const CODE_TABLE = new Map();
for (let f = MIN_FLOORS; f <= MAX_FLOORS; f++) {
  CODE_TABLE.set(promoCode(f), f);
}

/**
 * Validate a promo code and recover the floor count that earned it.
 * Tolerant of case, surrounding whitespace, and a missing "hu-" prefix.
 *
 * @param {string} input
 * @returns {{ valid: boolean, floors: number | null }}
 */
export function parsePromoCode(input) {
  if (typeof input !== "string") return { valid: false, floors: null };
  let code = input.trim().toLowerCase();
  if (!code.startsWith("hu-")) code = "hu-" + code;
  const floors = CODE_TABLE.get(code);
  return floors === undefined
    ? { valid: false, floors: null }
    : { valid: true, floors };
}

export { MIN_FLOORS, MAX_FLOORS };
